cmd/podman/tree.go

```diff
@@ -0,0 +1,194 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/containers/libpod/cmd/podman/libpodruntime"
+	"github.com/containers/libpod/libpod/image"
+	units "github.com/docker/go-units"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+//InfoImage keep information of Image along with all associated layers
+type InfoImage struct {
+	// Image ID
+	ID string
+	// Repo tags for image
+	RepoTags []string
+	// Size of Image calculated from Layer size.
+	Size int64
+	// Stores all layers of image.
+	Layers []image.LayerInfo
+}
+
+var (
+	treeFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "whatrequires",
+			Usage: "Show all child images in tree structure",
+		},
+	}
+	treeDescription = "Displays the image hierarchy in local image cache. The information is printed in tree format"
+
+	treeCommand = cli.Command{
+		Name:                   "tree",
+		Usage:                  "Show the image hierarchy in local image cache",
+		Description:            treeDescription,
+		Action:                 treeCmd,
+		Flags:                  treeFlags,
+		ArgsUsage:              "",
+		UseShortOptionHandling: true,
+	}
+)
+
+func treeCmd(c *cli.Context) error {
+
+	runtime, err := libpodruntime.GetRuntime(c)
+	if err != nil {
+		return errors.Wrapf(err, "could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	args := c.Args()
+	if len(args) == 0 {
+		return errors.Errorf("an image name must be specified")
+	}
+	if len(args) > 1 {
+		return errors.Errorf("podman tree takes at most 1 argument")
+	}
+
+	image, err := runtime.ImageRuntime().NewFromLocal(args[0])
+	if err != nil {
+		return err
+	}
+
+	// Fetch map of image-layers, which is used for printing output.
+	layerInfoMap, err := image.GetLayersMapWithImageInfo(context.Background())
+	if err != nil {
+		return errors.Wrapf(err, "error while retriving layers of image %q", image.InputName)
+	}
+
+	// TopLayer of image is used as start point for traversing in LayerMap.
+	imageTopLayer := image.TopLayer()
+	ll, ok := layerInfoMap[imageTopLayer]
```

Comment:
(Non-blocking: this is unnecessary AFAICS. `ll` is only used for `ll.ID` == `imageTopLayer`, so its uses could use `imageTopLayer` directly.)

Reply:
fixed